### PR TITLE
SCF-Session-5

### PR DESCRIPTION
**Pull request by meta-ishita-mathur**

### CODE REVIEW

File: SCF-Session-5/lcmhcf/LcmHcfTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	{
+		int lcmOutput = lcmHcfObject.lcm(24, 60);
+		assertEquals(120, lcmOutput);
+		
+		int hcfOutput = lcmHcfObject.hcf(24, 60);
+		assertEquals(12, hcfOutput);
+
+	}
+	
+	@org.junit.Test(expected = AssertionError.class)
```
Comment (sandeep-sky) [inline SCF-Session-5/lcmhcf/LcmHcfTest.java]: Why it is `@org.junit.Test` while others are just `@Test`?

File: SCF-Session-5/lcmhcf/LcmHcf.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package lcmhcf;
+
+/**
+ * This class contains methods to find LCM and HCF of two numbers using recursion.
+ * @author Ishita Mathur
+ *
+ */
+public class LcmHcf
```
Comment (sandeep-sky) [inline SCF-Session-5/lcmhcf/LcmHcf.java]: Class and method names `LcmHcf`, `lcm` & `hcf`.

File: SCF-Session-5/lcmhcf/LcmHcf.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
+package lcmhcf;
+
+/**
+ * This class contains methods to find LCM and HCF of two numbers using recursion.
+ * @author Ishita Mathur
+ *
+ */
+public class LcmHcf 
+{
+	static int i = 1;
```
Comment (sandeep-sky) [inline SCF-Session-5/lcmhcf/LcmHcf.java]: Why `static int i = 1` is declared at class level?

File: SCF-Session-5/lcmhcf/LcmHcfTest.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import static org.junit.Assert.*;
+
+import org.junit.Test;
+
+public class LcmHcfTest 
+{
+	LcmHcf lcmHcfObject = new LcmHcf();
+	
+	@Test
+	public void lcmTest()
```
Comment (sandeep-sky) [inline SCF-Session-5/lcmhcf/LcmHcfTest.java]: Method name is `lcmTest` but it has a test for `hcf` also. Name should be clear about the method's functionality and there should be different test methods for each method of the class.

File: SCF-Session-5/searches/Search.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	public int linearSearch(int inputArray[], int element)
+    	{
+		int upperLimit;
+		if (inputArray.length == 0)
+			throw new AssertionError("Array is empty");
+		
+		upperLimit = inputArray.length - 1;
+        
+		if (upperLimit < lowerLimit)
+        		return -1;
```
Comment (sandeep-sky) [inline SCF-Session-5/searches/Search.java]: Please suggest any use case for this if condition.

File: SCF-Session-5/searches/SearchTest.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import static org.junit.Assert.*;
+
+import org.junit.Test;
+
+public class SearchTest
+{
+	Search searchObject = new Search();
+	
+	@Test
+	public void testPositive()
```
Comment (sandeep-sky) [inline SCF-Session-5/searches/SearchTest.java]: Use separate test methods for each method in class and test method name should be relevant.

File: SCF-Session-5/searches/SearchTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		
+		int binaryIndex = searchObject.binarySearch(new int[]{1, 2, 3, 4, 5, 6}, 0, 5, 7);
+		assertEquals(-1, binaryIndex);	
+	}
+	
+	@Test(expected = AssertionError.class)
+	public void testLinearException()
+	{
+		int linearIndex = searchObject.linearSearch(new int[]{}, 0);
+		assertEquals(0, linearIndex);
```
Comment (sandeep-sky) [inline SCF-Session-5/searches/SearchTest.java]: Exception is not asserted anywhere in your test classes. Write a test method where exception can be asserted.

File: SCF-Session-5/searches/SearchTest.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	}
+	
+	@Test(expected = AssertionError.class)
+	public void testBinaryException()
+	{
+		int binaryIndex = searchObject.binarySearch(new int[]{}, 0, 0, 0);
+		assertEquals(0, binaryIndex);
+	}
+
```
Comment (sandeep-sky) [inline SCF-Session-5/searches/SearchTest.java]: Code block `else if(inputArray[middle] > element)`  of `binarySearch` is not covered in test methods. Why?

File: SCF-Session-5/N_Queens/NQueenTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package N_Queens;
+
+import static org.junit.Assert.*;
+
+import org.junit.Test;
```
Comment (sandeep-sky) [inline SCF-Session-5/N_Queens/NQueenTest.java]: `import org.junit.Test;` is never used in your test class, why it is imported?

File: SCF-Session-5/N_Queens/NQueenTest.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+public class NQueenTest
+{
+	NQueenProblem nQueen = new NQueenProblem();
+    
+    	/**
+     	* Test case for n = 4 (solution exists)
+     	*/
+    	@org.junit.Test
```
Comment (sandeep-sky) [inline SCF-Session-5/N_Queens/NQueenTest.java]: Why `@org.junit.Test` instead of `@test`?

File: SCF-Session-5/N_Queens/NQueenProblem.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    	{
+        	int row = 0; 
+        	int[][] board = new int[N][N];
+        
+        	// initializing the board with 0 at all places
+        	for(int i = 0; i < N; i++)
+            		for(int j = 0; j < N; j++)
+                		board[i][j] = 0;
+        
+        	boolean result = placeQueen(N, board, row);
```
Comment (sandeep-sky) [inline SCF-Session-5/N_Queens/NQueenProblem.java]: Variables should be declared at top of the any block. Statement can be written as `return placeQueen(N, board, row);`.

File: SCF-Session-5/N_Queens/NQueenProblem.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    
+    	/**
+     	* places the queens on the board at positions where they can be placed, if possible
+     	* @param N dimension of matrix
+     	* @param board
+     	* @param row
+     	* @return
+     	*/
+    	private boolean placeQueen(int N, int[][] board, int row)
+    	{
```
Comment (sandeep-sky) [inline SCF-Session-5/N_Queens/NQueenProblem.java]: Need to discuss the the functionality of this method. Please describe it in next meeting.

File: SCF-Session-5/lcmhcf/LcmHcf.java (hunk 3 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/lcmhcf/LcmHcf.java]: It was discussed in meeting

File: SCF-Session-5/lcmhcf/LcmHcf.java (hunk 2 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/lcmhcf/LcmHcf.java]: class name can't be changed now but I have updated the method names and will take care about this in future.

File: SCF-Session-5/lcmhcf/LcmHcfTest.java (hunk 1 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/lcmhcf/LcmHcfTest.java]: done

File: SCF-Session-5/lcmhcf/LcmHcfTest.java (hunk 4 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/lcmhcf/LcmHcfTest.java]: done

File: SCF-Session-5/searches/Search.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
+package searches;
+
+/**
+ * This class contains methods to perform linear and binary searches.
+ * @author Ishita_Mathur
+ *
+ */
+public class Search 
+{
+	int lowerLimit = 0;
```
[the remark this answers is not in the document]
Comment (meta-ishita-mathur) [reply SCF-Session-5/searches/Search.java]: It was discussed in meeting.

File: SCF-Session-5/searches/Search.java (hunk 5 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/searches/Search.java]: done

File: SCF-Session-5/searches/SearchTest.java (hunk 6 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/searches/SearchTest.java]: done

File: SCF-Session-5/searches/SearchTest.java (hunk 7 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/searches/SearchTest.java]: done

File: SCF-Session-5/searches/SearchTest.java (hunk 8 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/searches/SearchTest.java]: done

File: SCF-Session-5/N_Queens/NQueenProblem.java (hunk 12 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/N_Queens/NQueenProblem.java]: done

File: SCF-Session-5/N_Queens/NQueenTest.java (hunk 10 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/N_Queens/NQueenTest.java]: done

File: SCF-Session-5/N_Queens/NQueenProblem.java (hunk 11 above)
Comment (meta-ishita-mathur) [reply SCF-Session-5/N_Queens/NQueenProblem.java]: done

**Pull request merged**